Mathlib/Analysis/Calculus/SecondDerivativeTest.lean

```diff
@@ -0,0 +1,142 @@
+/-
+Copyright (c) 2024 Bjørn Kjos-Hanssen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bjørn Kjos-Hanssen
+-/
+import Mathlib.Analysis.Calculus.FirstDerivativeTest
+import Mathlib.Order.Interval.Set.Basic
+
+/-!
+# The Second-Derivative Test
+
+We prove the Second-Derivative test from calculus using the First-Derivative test.
+Source: [Wikipedia](https://en.wikipedia.org/wiki/Derivative_test#Proof_of_the_second-derivative_test).
+
+## Main results
+
+* `isLocalMin_of_deriv_deriv_pos`: The second-derivative test, minimum version.
+
+## Tags
+
+derivative test, calculus
+-/
+
+open Set Filter Topology
+
+
+/-- Insert a missing point between two adjacent open real intervals. -/
+theorem insert_Ioo₀ {x ε₀ ε₁ : ℝ} (hε₀ : ε₀ > 0) (hε₁ : ε₁ > 0):
+    insert x (Ioo (x - ε₀) x ∪ Ioo x (x + ε₁)) = Ioo (x - ε₀) (x + ε₁) :=
+  insert_Ioo ⟨by linarith,by linarith⟩
+
+
+section SecondDeriv
+
+variable {f : ℝ → ℝ} {x₀ : ℝ}
+
+/-- If the slope from a critical point `x₀` to `b > x₀` is positive then so is the derivative
+ at `b`. -/
+lemma slopeSimpPos {b : ℝ} (hb : x₀ < b) (hbf : 0 < slope (deriv f) x₀ b)
+    (hf : deriv f x₀ = 0) : 0 < deriv f b := by
+  unfold slope at hbf
+  rw [hf] at hbf
+  simp_all
+
+/-- If the slope from a critical point `x₀` to `b < x₀` is positive then the derivative
+ at `b` is negative. -/
+lemma slopeSimpNeg {b : ℝ} (hb : b < x₀) (hbf : 0 < slope (deriv f) x₀ b)
+    (hf : deriv f x₀ = 0) : deriv f b < 0 := by
+  unfold slope at hbf
+  simp_rw [smul_eq_mul, hf] at hbf
+  rw [mul_comm] at hbf
+  have : 0 < deriv f b / (b - x₀) := sub_zero (deriv f b) ▸ hbf
+  contrapose this
+  exact not_lt.mpr <| div_nonpos_of_nonneg_of_nonpos (not_lt.mp this) (by linarith)
+
+/-- If the derivative is negative (positive) to the left (right) then the
+function is differentiable in a punctured neighborhood. -/
+theorem eventually_differentiable_of_deriv_nonzero {ε : ℝ}
+    (hε : ε > 0)
+    (hε₀ : ∀ b ∈ Ioo (x₀ - ε) x₀, deriv f b < 0)
+    (hε₁ : ∀ b ∈ Ioo x₀ (x₀ + ε), 0 < deriv f b) :
+    ∀ᶠ x in 𝓝[≠] x₀, DifferentiableAt ℝ f x :=
```

Comment:
Is this `DifferentiableWithinAt`?

Reply:
Well, I just stated it in a way that seemed useful in order to prove the 2nd derivative test below. But now that you mention it, this lemma is really all about `ne` rather than `lt` and `gt` so I have restated the lemma in a new version.